src/Soql/SoqlConnection.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+namespace Codelicia\Soql;
+
+use Assert\Assertion;
+use Doctrine\DBAL\Driver\Connection;
+use Doctrine\DBAL\ParameterType;
+use GuzzleHttp\Client;
+use function func_get_args;
+use function json_decode;
+use function sprintf;
+
+class SoqlConnection implements Connection
+{
+    /** @var string */
+    private $salesforceInstance;
+
+    /** @var string */
+    private $consumerKey;
+
+    /** @var string */
+    private $consumerSecret;
+
+    /** @var string */
+    private $username;
+
+    /** @var string */
+    private $password;
+
+    /** @var Client */
+    private $conn;
+
+    public function __construct(array $params, string $username, string $password)
+    {
+        Assertion::keyExists($params, 'salesforceInstance');
+        Assertion::keyExists($params, 'consumerKey');
+        Assertion::keyExists($params, 'consumerSecret');
+
+        $this->salesforceInstance = $params['salesforceInstance'];
+        $this->consumerKey = $params['consumerKey'];
+        $this->consumerSecret = $params['consumerSecret'];
+
+        $this->username = $username;
+        $this->password = $password;
+
+        $token = $this->retrieveAccessToken();
```

Comment:
It seems like a named constructor would be a good choice here, then the constructor could receive a `ClientInterface` and be testable.

Reply:
I think that a participant responsible for retrieving access tokens would fit better here. Then I'd have this new participant as a dependency instead of `ClientInterface`.

I wonder if from one instance of `Connection` the access token gets expired, how would the client handle this? 

By having a class to retrieve access tokens could ease this process. A cache implementation on top could even be possible to avoid issuing one access token per request.

What do you guys think?

Reply:
I took into consideration both of your reviews and I've also discussed with @batusa in offline about the creation of the `Guzzle\HttpClient`. I've transferred the creation of it to the `SoqlDriver` and divided the `Client` into two interfaces, `AccessTokenFactory` and `AuthorizedClientFactory`.

Then if the user wants to replace it by something else, they can set it up in the configuration as in:
```php
$config [
    'authorizedClientFactory' => new AuthorizedClientFactory(
        new AccessTokenFactory(/*...*/)
    ),
];

/** @var SoqlConnection $conn */
$conn = DriverManager::getConnection($connectionParams, $config);
```